### PR TITLE
Added country search

### PR DESCRIPTION
**Pull request by fullstackctp**
Added the functionality for searching countries #867 and #871

### CODE REVIEW

File: lib/framework/data_context.dart [hunk 1]
```diff
@@ -543,6 +546,96 @@ class EnsembleStorage with Invokable {
...
 }
 
+/// singleton handling utils
+class EnsembleUtils with Invokable {
+  static final EnsembleUtils _obj = EnsembleUtils._();
+  EnsembleUtils._();
+  factory EnsembleUtils() => _obj;
+
+  @override
+  Map<String, Function> getters() => {'getCountries': () => allCountries};
```
Comment (vusters) [inline lib/framework/data_context.dart:556]: You want to move this to method so you can use `utils.getCountries()`. Getters syntax will be `utils.getCountries` which is not what we want.

File: lib/framework/data_context.dart [hunk 2]
```diff
@@ -543,6 +546,96 @@ class EnsembleStorage with Invokable {
...
+
+  @override
+  Map<String, Function> methods() => {
+        'getCountries': (value) {
```
Comment (vusters) [inline lib/framework/data_context.dart:560]: Please review the jiras requirements:
#867 - support `getCountries()` and `getCountry(<alpha-2>)`
#871 - support `findCountry(<search string>)`

File: lib/framework/data_context.dart [hunk 3]
```diff
@@ -543,6 +546,96 @@ class EnsembleStorage with Invokable {
...
+  @override
+  Map<String, Function> setters() => {};
+}
+
+class CountryManager {
+  static final CountryManager _instance = CountryManager._();
+  CountryManager._();
+  factory CountryManager() => _instance;
+
+  List<Map<String, dynamic>> getCountryData(String userInput) {
```
Comment (vusters) [inline lib/framework/data_context.dart:575]: Please simply this logic.

**Review by vusters — COMMENTED** [review]
Use this for search:
- always search for name if there's at least 1 character
- additonally, search for alpha-2 if at most 2 character, and search for alpha-3 if at most 3 character

File: lib/framework/data_context.dart [hunk 4]
```diff
@@ -546,45 +546,52 @@ class EnsembleStorage with Invokable {
...
   Map<String, Function> setters() => {};
-}
 
-class CountryManager {
-  static final CountryManager _instance = CountryManager._();
-  CountryManager._();
-  factory CountryManager() => _instance;
+  List<Map<String, dynamic>> getCountry(String val) {
+    String input = val.toLowerCase().trim();
+    List<Map<String, dynamic>> alpha2List = [];
```
Comment (vusters) [inline lib/framework/data_context.dart:571]: We shouldn't have to iterate through the list if the # of characters is not 2 - it's wasteful.

Even better, let's search by alpha-2 if 2 characters, and alpha-3 if 3 characters.

File: lib/framework/data_context.dart [hunk 5]
```diff
@@ -546,45 +546,52 @@ class EnsembleStorage with Invokable {
...
+      countryCode = countryCode.toLowerCase();
+      if (countryCode == input) {
+        alpha2List.add(i);
+      }
+    }
+    return alpha2List;
+  }
 
-  List<Map<String, dynamic>> getCountryData(String userInput) {
+  List<Map<String, dynamic>> findCountry(String userInput) {
```
Comment (vusters) [inline lib/framework/data_context.dart:582]: Problems with this function:
- you are making 2 passes through the list. We should only do it once for performance reason.
- By concating the list, you essentially break the existing alphabetical sort order. 

Please use 1 pass only and keep the original JSON sort order.

**Review by vusters — CHANGES_REQUESTED** [review]

File: lib/framework/data_context.dart [hunk 6]
```diff
@@ -569,74 +569,41 @@ class EnsembleUtils with Invokable {
   List<Map<String, dynamic>> getCountry(String val) {
     String input = val.toLowerCase().trim();
     List<Map<String, dynamic>> alpha2List = [];
-    for (var i in allCountries) {
-      String countryCode = i['iso']['alpha-2'];
-      countryCode = countryCode.toLowerCase();
-      if (countryCode == input) {
-        alpha2List.add(i);
+    if (input.length == 2) {
```
Comment (vusters) [inline lib/framework/data_context.dart:572]: please search alpha-2 for 2 characters, and alpha-3 for 3 characters.

File: lib/framework/data_context.dart [hunk 7]
```diff
@@ -569,74 +569,41 @@ class EnsembleUtils with Invokable {
...
+      for (var i in allCountries) {
+        String countryCode = i['iso']['alpha-2'];
+        countryCode = countryCode.toLowerCase();
+        if (countryCode == input) {
+          alpha2List.add(i);
+        }
```
Comment (vusters) [inline lib/framework/data_context.dart:578]: The value is unique. You need to break out and return once found. There is no reason to loop through the entire list.
The return type should be `Map<String, dynamic>?`. Either you find the single result or you don't.

**Review by vusters — CHANGES_REQUESTED** [review]

**Review by vusters — APPROVED** [review]

**Comment by kmahmood74** [comment]
@fullstackctp this is incomplete without a kitchen sink example and documentation. No one knows how to use it. Please provide example and documentation.

**Pull request merged**